flink/src/main/java/org/apache/iceberg/flink/source/reader/FlinkBulkFormatAdaptor.java

```diff
@@ -0,0 +1,222 @@
+    private Reader<T> currentReader;
+
+    ReaderAdaptor(
+        Map<FileFormat, BulkFormat<T, FileSourceSplit>> bulkFormatProvider,
+        Configuration config,
+        IcebergSourceSplit icebergSplit,
+        boolean isRestored) throws IOException {
+      this.config = config;
+      this.bulkFormatProvider = bulkFormatProvider;
+      this.fileIterator = icebergSplit.task().files().iterator();
+      this.isRestored = isRestored;
+
+      final CheckpointedPosition position = icebergSplit.checkpointedPosition();
+      if (position != null) {
+        // skip files based on offset in checkpointed position
+        Preconditions.checkArgument(position.getOffset() < icebergSplit.task().files().size(),
+            String.format("Checkpointed file offset is %d, while CombinedScanTask has %d files",
+                position.getOffset(), icebergSplit.task().files().size()));
+        for (int i = 0; i < position.getOffset(); ++i) {
+          fileIterator.next();
+          fileOffset++;
+        }
+        // first file may need to skip records
+        setupReader(position.getRecordsAfterOffset());
+      } else {
+        setupReader(0L);
+      }
+    }
+
+    /**
+     * TODO: we can't return RecordIterator with empty data.
+     * Otherwise, caller may assume it is end of input.
+     * We probably need to add a {@code hasNext()} API to
+     * {@link RecordIterator} to achieve the goal.
+     */
+    @Nullable
+    @Override
+    public RecordIterator<T> readBatch() throws IOException {
+      RecordIterator<T> result = null;
+      while (currentReader != null || fileIterator.hasNext()) {
+        if (currentReader == null) {
+          setupReader(0L);
+        } else {
+          result = currentReader.readBatch();
+          if (result != null) {
+            break;
+          } else {
+            closeCurrentReader();
+          }
+        }
+      }
+      if (result == null) {
+        return null;
+      } else {
+        return new RecordIteratorAdaptor(fileOffset, result);
+      }
+    }
+
+    @Override
+    public void close() throws IOException {
+      closeCurrentReader();
+    }
+
+    private void closeCurrentReader() throws IOException {
+      if (currentReader != null) {
+        currentReader.close();
+        currentReader = null;
+      }
+    }
+
+    private void setupReader(long skipRecordCount) throws IOException {
+      if (fileIterator.hasNext()) {
+        final FileScanTask fileScanTask = fileIterator.next();
+        final FileFormat fileFormat = fileScanTask.file().format();
+        if (!bulkFormatProvider.containsKey(fileFormat)) {
+          throw new IOException("Unsupported file format: " + fileFormat);
+        }
+        final BulkFormat<T, FileSourceSplit> bulkFormat = bulkFormatProvider.get(fileFormat);
+        fileOffset++;
+        final FileSourceSplit fileSourceSplit = new FileSourceSplit(
+            "",
+            new Path(URI.create(fileScanTask.file().path().toString())),
```

Comment:
So the  `FileSourceSplit` will use the flink's fs interface to access the underlying files ?   We iceberg currently has our own `FileIO` interface,  the object storage services are implementing this interface to write/read data to cloud.  If we introduce flink fs here ,  I'm concerning that we have to implement both flink fs interfaces and iceberg FileIO interfaces for making the experimental unified source work.

Reply:
We extended from `FileSourceSplit` mainly for the `BulkFormat` batch reader interface so that we can plug in vectorized readers from Flink. I am also debating if this is the right thing to do as mentioned in the original description.

But this is not really relevant to FileIO, which deals with underneath filesystem like S3. Here we are mainly talking about file format reader (like Parquet, Orc).